### PR TITLE
Add Google Kubernetes Engine Container Provider

### PR DESCRIPTION
**Pull request by agrare**
Screenshot from 2021-08-12 09-29-48

Screenshot from 2021-08-12 09-33-59

Dependent:
* https://github.com/ManageIQ/manageiq-decorators/pull/51
* https://github.com/ManageIQ/manageiq-documentation/pull/1588

Closes https://github.com/ManageIQ/manageiq-providers-google/issues/180

### CODE REVIEW

File: app/models/manageiq/providers/google/container_manager.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  def self.kubernetes_auth_options(options)
+    auth_options = {}
+    auth_options[:bearer_token] = google_access_token(options[:bearer])
+    auth_options
+  end
+
+  def self.google_access_token(google_json_key)
+    require "googleauth"
+    credentials = ::Google::Auth::ServiceAccountCredentials.make_creds(
```
Comment (agrare) [inline app/models/manageiq/providers/google/container_manager.rb:28]: NOTE: kubeclient does have a [`Kubeclient::GoogleApplicationDefaultCredentials`](https://github.com/abonas/kubeclient/blob/e03dd5508f01bd632cee21e788318d790eabe187/lib/kubeclient/google_application_default_credentials.rb) credential type, but it uses [`Google::Auth.get_application_default(scopes)`](https://github.com/abonas/kubeclient/blob/e03dd5508f01bd632cee21e788318d790eabe187/lib/kubeclient/google_application_default_credentials.rb#L5) which only pulls creds from environment variables and known paths.  This way allows us to use the json service account from the database easier.

File: app/models/manageiq/providers/google/container_manager.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      :json_key_io => StringIO.new(google_json_key),
```
Comment (agrare) [inline app/models/manageiq/providers/google/container_manager.rb:29]: NOTE: Investigate if I initialize `ServiceAccountCredentials` by just passing in the json string directly

File: app/models/manageiq/providers/google/container_manager.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    {
+      :bearer_token => google_access_token(options[:bearer])
+    }
+  end
+
+  def self.google_access_token(json_key)
+    require "googleauth"
+
+    client_email, private_key, project_id, quota_project_id =
+      JSON.parse(json_key).values_at("client_email", "private_key", "project_id", "quota_project_id")
```
Comment (Fryguy) [inline app/models/manageiq/providers/google/container_manager.rb:38]: If someone uploads junk as the json_key will this just blow up?  Should we be defensive either here or perhaps at the authentications level (or are we already)?

File: app/models/manageiq/providers/google/container_manager.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    private_key = OpenSSL::PKey::RSA.new(private_key)
```
Comment (Fryguy) [inline app/models/manageiq/providers/google/container_manager.rb:40]: Can the key only be rsa or could there be, say, ed25519 keys?

File: app/models/manageiq/providers/google/container_manager/refresh_worker/watch_thread.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    notice.object&.kind == "Endpoints" && filter_endpoint?(notice.object)
+  end
+
+  def filter_endpoint?(endpoint)
+    # The base kubernetes parser uses the endpoint subset addresses and targetRefs
+    # to build "container_groups_refs" in order to link pods to container_services
+    #
+    # If an endpoint doesn't have any subsets then it is a pointless update
+    endpoint.subsets.blank?
+  end
```
Comment (Fryguy) [inline app/models/manageiq/providers/google/container_manager/refresh_worker/watch_thread.rb:14]: Should this just belong in the base kubernetes code?  It seems the same argument in the comment would apply there as well, right?

File: app/models/manageiq/providers/google/container_manager.rb (hunk 4 above)
Comment (agrare) [reply app/models/manageiq/providers/google/container_manager.rb:40]: This comes from the google-auth library (see https://github.com/ManageIQ/manageiq-providers-google/pull/195#discussion_r687183246)

I wanted to see if we could bypass the weirdness of having to pass in an IO when we already have the string.  The extra logic that I had to copy doesn't seem worth it though

File: app/models/manageiq/providers/google/container_manager/refresh_worker/watch_thread.rb (hunk 5 above)
Comment (agrare) [reply app/models/manageiq/providers/google/container_manager/refresh_worker/watch_thread.rb:14]: Yeah maybe?  I'll put in a PR to move it

**Comment by agrare** [comment]
Okay went back to using `.make_creds()` and dropped the endpoint notice filter from the watch_thread.  I'll put in a PR to add that to kubernetes but that isn't a dependency for this PR to make it in, just reduces some "useless" notices.

Tested streaming refresh with the watch notices and it all works as expected.  Taking out of WIP

@Fryguy PTAL

File: app/models/manageiq/providers/google/container_manager/refresh_worker/watch_thread.rb (hunk 5 above)
Comment (agrare) [reply app/models/manageiq/providers/google/container_manager/refresh_worker/watch_thread.rb]: https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/443

**Comment by Fryguy** [comment]
Backported to `morphy` in commit 69b5ac081e610dced3371a9481adac6f9e4be50b.

```text
commit 69b5ac081e610dced3371a9481adac6f9e4be50b
Author: Jason Frey <fryguy9@gmail.com>
Date:   Fri Aug 13 13:02:38 2021 -0400

    Merge pull request #195 from agrare/add_gke_provider
    
    Add Google Kubernetes Engine Container Provider
    
    (cherry picked from commit 0669121ae56d0c0c0b670916e915f0d811c28358)
```

**Pull request merged**